\documentclass{article}
\usepackage{graphicx} % Required for inserting images
\usepackage{authblk}
\usepackage{biblatex}
\usepackage{hyperref}
\emergencystretch=2.5pt
\hypersetup{
    colorlinks=true,    
    urlcolor=blue,
    pdftitle={Overleaf Example},
    pdfpagemode=FullScreen,
    citecolor=black
    }
\usepackage{geometry}
\usepackage{float}
\usepackage{comment}
\usepackage{amsmath}

\addbibresource{references.bib}

\title{MANUDB: database and application to retrieve, visualize and predict mammalian NUMTs}

\begin{document}
\author[1,2]{Bálint Biró}
\author[2]{Zoltán Gál}
\author[2]{Zsófia Nagy}
\author[2]{Orsolya Ivett Hoffmann}

\affil[1]{Group BM, Data Insights Team, \_VOIS, Kerepesi str. 35, 1087, Budapest, Hungary}
\affil[2]{Agribiotechnology and Precision Breeding for Food Security National Laboratory, Department of Animal Biotechnology, Institute of Genetics and Biotechnology, Hungarian University of Agriculture and Life Sciences, Szent-Györgyi Albert str. 4, 2100, Gödöllő, Hungary}
\date{}

\maketitle

\section{Abstract}
\section{Introduction}
\section{Materials and methods}
\subsection{Data collection and organization}
The data deposited in MANUDB was constructed using three sources that are associated with the National Center for Biotechnology Information - Reference sequence database (NCBI-RefSeq) \cite{ncbi_refseq}. Nuclear and mitochondrial genome sequences were downloaded using the public FTP sites of NCBI (\url{https://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/} and \url{https://ftp.ncbi.nlm.nih.gov/genomes/refseq/mitochondrion/}). Taxonomy related data was also gathered from public NCBI FTP site (\url{https://ftp.ncbi.nlm.nih.gov/genomes/GENOME_REPORTS/eukaryotes.txt}). For the sequence alignments between the nuclear genomes and their corresponding mitochondrial genomes LASTAL \cite{lastal} was used with the settings and e-value threshold as they were previously published \cite{tsuji}. After the alignment for each item e value, genomic id, genomic start, mitochondrial start, genomic length, mitochondrial length, genomic strand, mitochondrial strand, genomic sequence, mitochondrial sequence information were extracted. This data was further enhanced with order, family and genus taxonomic information.
\subsection{Database implementation}
To facilitate future maintenance and provide a robust way of accessing the data MANUDB is based on the SQLite relational database management system \cite{sqlite} and its corresponding database engine. The user interface was programmed within the Streamlit ecosystem using additional HTML and CSS to further improve user experience. Back-end and front-end is connected through Python scripts. The whole application is deployed within the Streamlit cloud.
\subsection{Functionalities}
MANUDB currently supports three functionalities namely retrieval, visualization and prediction.\\Using the retrieval functionality users are able to download specific datasets into .csv format files. The fields to retrieve can be selected based on the specific tables. Another option is to export the whole dataset with all the fields defined previously based on the specified species.\\By interacting with the visualization functionality users can generate and download publication ready figures. These chord diagrams display the relationships between the mitochondrial genome and different parts (chromosomes and scaffolds) of the nuclear genome. The visualization itself is performed by using the Python implementation of Circos \cite{circos,pycirclize}.\\For the prediction NUMTs were the positive labelled items. For negative labelled item generation a given genomic part was randomly sampled as many times as many NUMTs were located on the given genomic part. For instance, if the X chromosome of the mouse genome contains 2 NUMTs in a length of 123 and 304 bps, then this chromosome is going to be sampled two times at random positions in the same lengths i.e., 123 and 304 bps, respectively. As a next step Kmers (k=3) were extracted from the sequences. These features were evaluated using forward selection and area under the curve (AUROC) as metric. To overcome the cold start problem (i.e. include an irrelevant feature into the dataset as the first feature) of the forward selection the feature with the highest mutual information score (MIS) was picked as the first feature. To find the best suited model to the problem four classification algorithms were tested. We have investigated a simple tree based method (decision tree), a linear model (support vector machine), an ensembl model (xgboost) and the model that was previously used to classify NUMTs (random forest) \cite{biro2024mitochondrial}. To select the model we have used the combined version of the most commonly used metrics for binary classification problems \cite{metrics} namely AUROC, precision, F1 score, recall and accuracy. To combine these metrics the area of the corresponding polygon was calculated as follows:\\
\begin{equation}\label{eq:polygon_area}
    \begin{split}
        Ap & =\frac{1}{2}\sum_{i=1}^nm_i m_{i+1} \sin(\alpha)
    \end{split}
\end{equation}
where
\begin{itemize}
    \item[] ~$Ap$ is the area of the polygon
    \item[] ~$n$ is the number of metrics
    \item[] ~$m$ is the metrics
    \item[] ~$\alpha$ is $\frac{360}{n}$
\end{itemize}

The selected model was optimized using the random search algorithm with 5000 iterations and cross validation (k=10). Model selection was performed within the ScikitLearn ecosystem \cite{scikitlearn}. The tuned model works in the background of the prediction functionality.

\section{Results}

\begin{comment}
\begin{figure}[H]
    \centering
    \includegraphics[width=.5\textwidth]{ml_plot.png}
    \caption{Feature selection, model selection and optimization.}
    \label{fig:feature_selection}
\end{figure}

\begin{figure}[H]
    \centering
    \includegraphics[width=.75\textwidth]{MANUDB.png}
    \caption{The architecture of MANUDB}
    \label{fig:architecture}
\end{figure}

\begin{figure}[H]
    \centering
    \includegraphics[width=.5\textwidth]{chord.png}
    \caption{Sample chord diagrams}
    \label{fig:chords}
\end{figure}

\begin{figure}[H]
    \centering
    \includegraphics[width=.5\textwidth]{taxonomical_structure.png}
    \caption{The taxonomical structure of MANUDB}
    \label{fig:taxonomy}
\end{figure}
\end{comment}


\section{Discussion}
\section{Conclusions}
\section{References}
\printbibliography[heading=none]

\end{document}
